plugins/vppv2/puntplugin/README.md

```diff
@@ -1,29 +1,149 @@
 # Punt plugin
 
-Punt plugin defines [model](../model/punt/punt.proto) and supports three scenarios:
+**Written for: v1.7-vpp18.10**
 
-1. **IP punt redirect:** packet matching one of the VPP IP addresses but otherwise would be dropped will be send to 
-the defined next hop IP address via TX interface. The rule can be enhanced and input traffic filtered with RX 
-interface - redirect applies only to traffic received via RX.
+The punt plugin provides several options for how to configure the VPP to allow a specific IP traffic to be punted to the host TCP/IP stack. The plugin supports **punt to the host** (either directly, or **via Unix domain socket**) and registration of **IP punt redirect** rules.
 
-2. **Punt to host:** packet matching one of the VPP IP address and also defined L3 protocol, L4 protocol and also port
-is punted to host
+- [Punt to host stack](#pths)
+  * [Model](#pths-model)
+  * [Configuration](#pths-config)
+  * [Limitations](#pths-limit)
+  * [Known issues](#pths-issues)
+- [IP redirect](ipr)
+  * [Model](#ipr-model)
+  * [Configuration](#ipr-config)
+  * [Limitations](#ipr-limit)
+  * [Known issues](#ipr-issues)
 
-3. **Punt to host via socket** packet has to match all the criteria as in previous case, but packet is punted to 
-unix domain socket
+## <a name="pths">Punt to host stack</a>
 
-In order to use punt to socket, VPP has to be started with configuration:
+All the incoming traffic matching one of the VPP interface addresses, and also matching defined L3 protocol, L4 protocol, and port - and would be otherwise dropped - will be instead punted to the host. If a Unix domain socket path is defined (optional), the traffic will be punted via socket. All the fields which serve as a traffic filter are mandatory.
+
+### <a name="pths-model">Model</a>
+
+The punt plugin defines the following [model](../model/punt/punt.proto) which grants support for two main configuration items defined by different northbound keys.
+
+The punt to host is defined as `ToHost` object in the generated proto model. 
+
+The key has a following format (for both, with or without socket):
+```
+vpp/config/v2/punt/tohost/l3/<L3_protocol>/l4/<L4_protocol>/port/<port_number>
+```
+
+L3/L4 protocol in the key is defined as a `string` value, however, the value is transformed to numeric representation in the VPP binary API. To clarify key values, here is a table of supported protocols:
+
+The usage of L3 protocol `ALL` is exclusive for IP punt to host (without socket registration) in the VPP API. If used for the IP punt with socket registration, the vpp-agent calls the binary API twice with the same parameters for both, IPv4 and IPv6. 
+
+**Note:** labels `UNDEFINED_L3` and `UNDEFINED_L4` should not be used on purpose since it may cause errors in the vpp-agent or in the VPP.
```

Comment:
This is not needed now, since the previous mentions were removed.

Reply:
This has nothing to do with previous mentions. Both labels are valid and can be found in the model so I consider important to mention that they should not be supposed to use in configuration.

Reply:
The implication that those values should not be used should be very clear to users after reading the name itself (`UNDEFINED_`).
And even if someone actually used it, then the value would simply be ignored, because it equals the default zero value. 
For these reasons I think mentioning it in README has no real value and just bloats the documentation.

Reply:
> And even if someone actually used it, then the value would simply be ignored, because it equals the default zero value.

This is not true, values which can be put as the L3/L4 proto are strictly defined by the VPP (that was the reason why I put those tables here). The zero value causes non-repeatable error in the vpp-agent (already mentioned in that line) or, if passed to the vpp it returns some unspecified error (error code -1 if I remeber correctly).
Whatever, I'll remove it.